File: testdata/fixtures/java/CWE-434b.java
// CWE-434b: Unrestricted upload second — getSubmittedFileName reaches Files.write/new File
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import javax.servlet.http.Part;

public class CWE434b {
    public void vuln1(Part filePart) throws Exception {
        String fileName = filePart.getSubmittedFileName();
        // ruleid: java-unrestricted-upload
        Files.write(Path.of("/uploads/" + fileName), filePart.getInputStream().readAllBytes());
    }

    public void vuln2(Part part) throws Exception {
        String name = part.getSubmittedFileName();
        // ruleid: java-unrestricted-upload
        File f = new File("/tmp/" + name);
    }

    public void vuln3(javax.servlet.http.HttpServletRequest request) throws Exception {
        Part p = request.getPart("file");
        String n = p.getSubmittedFileName();
        // ruleid: java-unrestricted-upload
        new File("/uploads", n);
    }

    public void safeFixed() throws Exception {
        String fileName = "fixed.txt";
        // ok: java-unrestricted-upload
        Files.write(Path.of("/uploads/" + fileName), "fixed".getBytes());
        // ok: java-unrestricted-upload
        File f = new File("/tmp/" + fileName);
    }

    public void safeWhitelist(Part filePart) throws Exception {
        String fileName = filePart.getSubmittedFileName();
        if (!fileName.endsWith(".png")) throw new IllegalArgumentException("bad");
        // ok: java-unrestricted-upload
        Files.write(Path.of("/uploads/safe.png"), filePart.getInputStream().readAllBytes());
    }
}

File: testdata/fixtures/c/CWE-434b.c
/* CWE-434b: fopen/userFilename without basename — unrestricted upload */
#include <stdio.h>
#include <stdlib.h>
#include <libgen.h>

void upload_vuln1(char *userFilename) {
    // ruleid: c-434-upload
    FILE *f = fopen(userFilename, "w");
    (void)f;
}

void upload_vuln2(int argc, char **argv) {
    char *name = argv[1];
    // ruleid: c-434-upload
    FILE *g = fopen(name, "w");
    (void)g;
}

void upload_vuln3() {
    char *userFilename = getenv("UPLOAD_NAME");
    // ruleid: c-434-upload
    FILE *h = fopen(userFilename, "wb");
    (void)h;
}

void upload_safe_basename(char *userFilename) {
    // ok: c-434-upload
    FILE *f = fopen(basename(userFilename), "w");
    (void)f;
}

void upload_safe_fixed() {
    // ok: c-434-upload
    FILE *f = fopen("/tmp/fixed.png", "w");
    (void)f;
}
